internal/controller/management/serviceaccounts/serviceaccounts_test.go

```diff
@@ -0,0 +1,117 @@
+	testCfg := ReconcilerConfig{
+		KargoNamespace: "fake-ns",
+	}
+	kubeClient := fake.NewClientBuilder().Build()
+	r := newReconciler(kubeClient, testCfg)
+	require.Equal(t, testCfg, r.cfg)
+	require.NotNil(t, kubeClient, r.client)
+
+	require.NotNil(t, r.getServiceAccountFn)
+	require.NotNil(t, r.updateServiceAccountFn)
+	require.NotNil(t, r.createRoleBindingFn)
+	require.NotNil(t, r.updateRoleBindingFn)
+	require.NotNil(t, r.deleteRoleBindingFn)
+	require.NotNil(t, r.listProjectFn)
+	require.NotNil(t, r.ensureControllerPermissionsFn)
+	require.NotNil(t, r.removeControllerPermissionsFn)
+}
+
+func TestEnsureControllerPermissions(t *testing.T) {
+	saName := "test-sa"
+	saNamespace := "default"
+	projectName := "project-1"
+
+	project := &kargoapi.Project{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+	serviceAccount := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      saName,
+			Namespace: saNamespace,
+		},
+	}
+
+	// Fake client with ServiceAccount and Project.
+	scheme := runtime.NewScheme()
+	_ = kargoapi.AddToScheme(scheme)
+	_ = corev1.AddToScheme(scheme)
+	_ = rbacv1.AddToScheme(scheme)
+	kubeClient := fake.NewClientBuilder().
+		WithScheme(scheme).
+		WithObjects(serviceAccount, project).
+		Build()
+
+	r := newReconciler(kubeClient, ReconcilerConfig{KargoNamespace: saNamespace})
+
+	err := r.ensureControllerPermissions(context.Background(), serviceAccount)
+	require.NoError(t, err)
+
+	roleBinding := &rbacv1.RoleBinding{}
+	err = kubeClient.Get(context.Background(), types.NamespacedName{
+		Name:      saName + "-readonly-secrets",
+		Namespace: projectName,
+	}, roleBinding)
+	require.NoError(t, err)
+	require.Equal(t, saName+"-readonly-secrets", roleBinding.Name)
+	require.Equal(t, "kargo-controller-secrets-readonly", roleBinding.RoleRef.Name)
+	require.Equal(t, saName, roleBinding.Subjects[0].Name)
+}
+
+func TestRemoveControllerPermissions(t *testing.T) {
+	saName := "test-sa"
+	projectName := "project-1"
+	roleBindingName := saName + "-readonly-secrets"
+
+	roleBinding := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: projectName,
+		},
+	}
+	project := &kargoapi.Project{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: projectName,
+		},
+	}
+
+	scheme := runtime.NewScheme()
+	_ = kargoapi.AddToScheme(scheme)
+	_ = rbacv1.AddToScheme(scheme)
+	kubeClient := fake.NewClientBuilder().
+		WithScheme(scheme).
+		WithObjects(roleBinding, project).
+		Build()
+
+	r := newReconciler(kubeClient, ReconcilerConfig{KargoNamespace: "fake-ns"})
+
+	err := r.removeControllerPermissions(context.Background(), types.NamespacedName{Name: saName})
+	require.NoError(t, err)
+
+	err = kubeClient.Get(context.Background(), types.NamespacedName{
+		Name:      roleBindingName,
+		Namespace: projectName,
+	}, roleBinding)
+	require.Error(t, err)
+	require.True(t, apierrors.IsNotFound(err), "expected NotFound error but got: %v", err)
```

Comment:
fwiw, we don't often bother with the last argument to require/assert functions, which provides an English description of the failure. It's seldom any more informative than if it were omitted. I don't mind it here, but I figure knowing this may save you some keystrokes in the future.